File: py/cnice/farewell.py
"""Locale-correct valedictions with explicit override support."""

from ._tables import TABLES

_TABLE = TABLES["closing"]
_KEYS = {key.lower(): key for key in _TABLE["locales"]}
__all__ = ["closing", "available_locales"]


def closing(locale: str, override_closing: str | None = None) -> str:
    """Resolve exact locale, base language, then English; an override wins."""
    if override_closing is not None:
        return override_closing
    lowered = locale.lower()
    key = _KEYS.get(lowered, _KEYS.get(lowered.split("-", 1)[0], _TABLE["fallback"]))
    return _TABLE["locales"][key]


def available_locales() -> list[str]:
    """Sorted canonical BCP 47 codes in the valediction table."""
    return sorted(_TABLE["locales"])

File: js/@corbet-labs/cnice/src/farewell.ts
/**
 * Deterministic locale-correct valedictions for formal correspondence.
 *
 * Pure TypeScript port of cnice::farewell (previously the cfarewell crate): zero dependencies, zero
 * Node APIs, synchronous, no I/O. Behavior is defined by `tables/*.json`
 * at the repository root; `tests/vectors/*.json` is the shared conformance
 * suite.
 */
import { CLOSING_TABLE } from './generated/farewell-tables.ts';

export interface ClosingTable {
    locales: Record<string, string>;
    fallback: string;
}

const table = CLOSING_TABLE as ClosingTable;

function baseLanguage(locale: string): string {
    return locale.split('-')[0] ?? locale;
}

function resolveKey(locale: string): string {
    const lower = locale.toLowerCase();
    if (Object.hasOwn(table.locales, lower)) return lower;
    const base = baseLanguage(lower);
    return Object.hasOwn(table.locales, base) ? base : table.fallback;
}

/**
 * Valediction for a BCP 47 locale. Unknown locales fall back through the
 * base language to English; an explicit override always wins (used for
 * per-workspace closing choices).
 */
export function closing(locale: string, overrideClosing?: string): string {
    if (overrideClosing !== undefined) return overrideClosing;
    return table.locales[resolveKey(locale)] ?? '';
}

/** BCP 47 locale codes with a valediction entry, sorted. */
export function availableLocales(): string[] {
    return Object.keys(table.locales).sort();
}
